0_From_Code_First_To_First_Principles/1_code_first/assignment/src/libs/utils/result.ts

```diff
@@ -0,0 +1,146 @@
+import { errorToString } from "./error-to-string";
+
+function isPromise<SuccessType>(value: any): value is Promise<SuccessType> {
+  return !!value && typeof value.then === "function";
+}
+
+// Source: https://khalilstemmler.com/articles/enterprise-typescript-nodejs/handling-errors-result-class/
+// + iterations thanks to ChatGPT and experiments
+
+export class Result<SuccessType, ErrorType = any> {
+  public isSuccess: boolean;
+  public isFailure: boolean;
+  public error?: string;
+  private _errorType?: ErrorType;
+  protected _value?: SuccessType;
+
+  protected constructor(
+    isSuccess: boolean,
+    error?: string,
+    errorType?: ErrorType,
+    value?: SuccessType
```

Comment:
What is the point of the `SuccessType` type? Why is it not simply `value`?

Reply:
It's a generic type parameter. Any function that returns a `Result` uses it to specify what it returns when successful.

Reply:
For example useful in `create-user.controller.ts` to see what the `CreateUser` use case returns: a result which either contains a `UserEntity` (ok case) or a `UserRepositoryError` (fail case):

<img width="648" alt="CleanShot 2023-10-03 at 19 07 14@2x" src="https://github.com/ericbolikowski/essentialism-ass1-user-api-code-first-approach/assets/10571311/e91e47de-4b29-49b8-96fc-c98c7345ede3">